src/projection.jl

```diff
@@ -187,7 +187,11 @@ function ProjectTo(xs::AbstractArray)
 end
 
 function (project::ProjectTo{AbstractArray})(dx::AbstractArray{S,M}) where {S,M}
-    # First deal with shape. The rule is that we reshape to add or remove trivial dimensions
+    # Trivial case:
+    if all(el -> el isa AbstractZero, dx)
+        return NoTangent()
```

Comment:
In the tests, this breaks `isempty((ProjectTo([]))(1:0))`. But what should happen?

Empty arrays can't really have a gradient, but on the other hand, sending them to `NoTangent()` is type-unstable.

Reply:
If it is a empy array of Any.
Then I think we can say the author of the code never cared about type stability anyway,
so we can send it to `NoTangent`, if we want.
But I guess this also breaks
`isempty((ProjectTo(Vector{Float64}[]))(1:0)` for arrays of arrays.
which we probably don't want to do, so we coukd insert a `!isempty` ?

Maybe we want to actually guard it by checking the eltype.
If the eltype is concrete we don't need to do the `all` check anyway?

SOmething for a follow up PR i guess

Reply:
When the eltype is abstract, I agree it's fine not to care about type stability, that battle is already lost.

Applying this `any` check with a concrete type was a bad idea, as it catches concrete empty, and thus destroys type-stabitlity for non-empty too. I just thought I was being clever and saving a method, since the concrete `fill(NoTangent())` would hit this.

Whether it's worth bothering about `Any[NoTangent()]` at all, maybe it's more clutter than it's worth?